### PR TITLE
Fix lowering of i64.extendN_s instructions

### PR DESCRIPTION
**Pull request by taylordotfish**
The i64-to-i32-lowering pass used by wasm2js doesn't handle i64.extend*N*_s instructions, leading to errors. This PR adds code to handle those instructions correctly.

### CODE REVIEW

File: test/wasm2js/sign_ext.wast [hunk 1]
```diff
@@ -5,4 +5,10 @@
   (func "test16" (param $x i32) (result i32)
     (i32.extend16_s (local.get $x))
   )
+  (func "test8_i64" (param $x i64) (result i64)
+    (i64.extend8_s (local.get $x))
+  )
+  (func "test16_i64" (param $x i64) (result i64)
+    (i64.extend16_s (local.get $x))
+  )
```
Comment (kripken) [inline test/wasm2js/sign_ext.wast:13]: I think we also want a test for `ExtendS32Int64`, the 32-bit case?

**Review by kripken — COMMENTED** [review]
Thanks! Looks correct, just one comment.

File: test/wasm2js/sign_ext.wast (hunk 1 above)
Comment (taylordotfish) [reply test/wasm2js/sign_ext.wast:13]: Done.

**Review by kripken — APPROVED** [review]
Thanks! Looks good.

Have you run the fuzzer on this? That basically means running `./scripts/fuzz_opt.py` for a few thousand iterations, just to see if it can't find some bug in this code.

**Comment by taylordotfish** [comment]
@kripken I'm getting consistent errors when I run `fuzz_opt.py`, including on the commit before this PR and on the current state of `main` (`03dca9aa1a5a266c53a474aeb3c10a3f0584b25b`).

This is with the default options and answering `y` to the initial contents `fuzz_opt.py` proposes—maybe I'm not passing in the right configuration?

Log:

```
warning: no output file specified, not emitting output
wabt found, but not wasm2c support dir
wabt found, but not wasm2c support dir
/home/taylor/binaryen/bin/wasm-opt --print-features /home/taylor/binaryen/test/hello_world.wat --all-features
POSSIBLE_FEATURE_OPTS: ['--disable-threads', '--disable-mutable-globals', '--disable-nontrapping-float-to-int', '--disable-simd', '--disable-bulk-memory', '--disable-sign-ext', '--disable-exception-handling', '--disable-tail-call', '--disable-reference-types', '--disable-multivalue', '--disable-gc', '--disable-memory64', '--disable-typed-function-references', '--disable-relaxed-simd']
checking infinite random inputs
- Perenially-important initial contents:
  lit/passes/optimize-instructions.wast
  passes/optimize-instructions_fuzz-exec.wast

- Recently added or modified initial contents (manually selected):
  lit/passes/once-reduction.wast
  passes/remove-unused-brs_enable-multivalue.wast
  lit/passes/optimize-instructions-bulk-memory.wast
  lit/passes/optimize-instructions-ignore-traps.wast
  lit/passes/optimize-instructions-gc.wast
  lit/passes/optimize-instructions-gc-iit.wast
  lit/passes/optimize-instructions-call_ref.wast
  lit/passes/inlining_splitting.wast
  heap-types.wast

Do you want to proceed with these initial contents? (Y/n) y
warning: cannot de-nan outside of function context
warning: cannot de-nan outside of function context
warning: cannot de-nan outside of function context
warning: cannot de-nan outside of function context
warning: cannot de-nan outside of function context
warning: no output file specified, not emitting output
warning: cannot de-nan outside of function context
warning: cannot de-nan outside of function context
warning: cannot de-nan outside of function context
warning: cannot de-nan outside of function context
warning: cannot de-nan outside of function context
warning: cannot de-nan outside of function context
warning: no output file specified, not emitting output
warning: no output file specified, not emitting output
warning: no output file specified, not emitting output
warning: cannot de-nan outside of function context
warning: cannot de-nan outside of function context
warning: no output file specified, not emitting output
warning: cannot de-nan outside of function context
warning: cannot de-nan outside of function context
warning: no output file specified, not emitting output

ITERATION: 1 seed: 4086517355219089600 size: 12577 (mean: 12577.0, stddev: 0.0) speed: 9058.97192224622 iters/sec,  0.0 wasm_bytes/sec

randomized pass debug: 
randomized feature opts: --all-features --disable-nontrapping-float-to-int --disable-bulk-memory --disable-multivalue --disable-memory64
randomized settings (NaNs, OOB, legalize): False False True
initial contents: /home/taylor/binaryen/test/spec/align64.wast
  picked submodule 12 from multi-module wast
(initial contents not valid for features, ignoring)
randomized opts: --heap2local --licm --remove-unused-module-elements --roundtrip --ssa --remove-unused-names

/home/taylor/binaryen/bin/wasm-opt input.dat -ttf -o a.wasm --denan --no-fuzz-oob --legalize-js-interface --all-features --disable-nontrapping-float-to-int --disable-bulk-memory --disable-multivalue --disable-memory64 --disable-exception-handling --disable-memory64 --disable-multivalue --strip-dwarf
pre wasm size: 12355
/home/taylor/binaryen/bin/wasm-opt a.wasm --all-features --disable-nontrapping-float-to-int --disable-bulk-memory --disable-multivalue --disable-memory64 --disable-exception-handling --disable-memory64 --disable-multivalue --strip-dwarf --print-features
['--enable-threads', '--enable-mutable-globals', '--enable-simd', '--enable-sign-ext', '--enable-tail-call', '--enable-reference-types', '--enable-gc', '--enable-typed-function-references', '--enable-relaxed-simd'] ['--strip-dwarf']
/home/taylor/binaryen/bin/wasm-opt a.wasm -o b.wasm --heap2local --licm --remove-unused-module-elements --roundtrip --ssa --remove-unused-names --denan --no-fuzz-oob --legalize-js-interface --enable-threads --enable-mutable-globals --enable-simd --enable-sign-ext --enable-tail-call --enable-reference-types --enable-gc --enable-typed-function-references --enable-relaxed-simd --strip-dwarf
post wasm size: 13286
running testcase handler: FuzzExec
/home/taylor/binaryen/bin/wasm-opt a.wasm --heap2local --licm --remove-unused-module-elements --roundtrip --ssa --remove-unused-names --enable-threads --enable-mutable-globals --enable-simd --enable-sign-ext --enable-tail-call --enable-reference-types --enable-gc --enable-typed-function-references --enable-relaxed-simd --strip-dwarf --fuzz-exec

running testcase handler: CheckDeterminism
/home/taylor/binaryen/bin/wasm-opt a.wasm -o b1.wasm --heap2local --licm --remove-unused-module-elements --roundtrip --ssa --remove-unused-names --enable-threads --enable-mutable-globals --enable-simd --enable-sign-ext --enable-tail-call --enable-reference-types --enable-gc --enable-typed-function-references --enable-relaxed-simd --strip-dwarf
/home/taylor/binaryen/bin/wasm-opt a.wasm -o b2.wasm --heap2local --licm --remove-unused-module-elements --roundtrip --ssa --remove-unused-names --enable-threads --enable-mutable-globals --enable-simd --enable-sign-ext --enable-tail-call --enable-reference-types --enable-gc --enable-typed-function-references --enable-relaxed-simd --strip-dwarf


Invocations so far:
   FuzzExec: 1
   CompareVMs: 0
   CheckDeterminism: 1
   Wasm2JS: 0
   Asyncify: 0

ITERATION: 2 seed: 307590667303464651 size: 4990 (mean: 8783.5, stddev: 3793.5) speed: 6.537124403651422 iters/sec,  83809.20341701306 wasm_bytes/sec

randomized pass debug: 
randomized feature opts: --all-features --disable-threads --disable-nontrapping-float-to-int --disable-simd --disable-bulk-memory --disable-tail-call --disable-reference-types --disable-exception-handling --disable-gc --disable-gc --disable-memory64 --disable-typed-function-references --disable-relaxed-simd
randomized settings (NaNs, OOB, legalize): True True True
randomized opts: --merge-locals --precompute-propagate --shrink-level=0

/home/taylor/binaryen/bin/wasm-opt input.dat -ttf -o a.wasm --legalize-js-interface --all-features --disable-threads --disable-nontrapping-float-to-int --disable-simd --disable-bulk-memory --disable-tail-call --disable-reference-types --disable-exception-handling --disable-gc --disable-gc --disable-memory64 --disable-typed-function-references --disable-relaxed-simd
pre wasm size: 2319
/home/taylor/binaryen/bin/wasm-opt a.wasm --all-features --disable-threads --disable-nontrapping-float-to-int --disable-simd --disable-bulk-memory --disable-tail-call --disable-reference-types --disable-exception-handling --disable-gc --disable-gc --disable-memory64 --disable-typed-function-references --disable-relaxed-simd --print-features
['--enable-mutable-globals', '--enable-sign-ext', '--enable-multivalue'] []
/home/taylor/binaryen/bin/wasm-opt a.wasm -o b.wasm --merge-locals --precompute-propagate --shrink-level=0 --legalize-js-interface --enable-mutable-globals --enable-sign-ext --enable-multivalue
post wasm size: 2727
running testcase handler: FuzzExec
/home/taylor/binaryen/bin/wasm-opt a.wasm --merge-locals --precompute-propagate --shrink-level=0 --enable-mutable-globals --enable-sign-ext --enable-multivalue --fuzz-exec


Invocations so far:
   FuzzExec: 2
   CompareVMs: 0
   CheckDeterminism: 1
   Wasm2JS: 0
   Asyncify: 0

ITERATION: 3 seed: 1862718566538268453 size: 91116 (mean: 36227.666666666664, stddev: 38935.30954744749) speed: 7.43823027640404 iters/sec,  76085.65749733693 wasm_bytes/sec

randomized pass debug: 
randomized feature opts: --all-features
randomized settings (NaNs, OOB, legalize): False False True
avoiding --flatten due to GC not supporting it (spilling of RTTs)
randomized opts: --coalesce-locals --vacuum --roundtrip --optimize-instructions --simplify-locals-notee-nostructure --rse --optimize-instructions --pick-load-signs -fimfs=99999999

/home/taylor/binaryen/bin/wasm-opt input.dat -ttf -o a.wasm --denan --no-fuzz-oob --legalize-js-interface --all-features
pre wasm size: 77017
/home/taylor/binaryen/bin/wasm-opt a.wasm --all-features --print-features
['--enable-threads', '--enable-mutable-globals', '--enable-nontrapping-float-to-int', '--enable-simd', '--enable-bulk-memory', '--enable-sign-ext', '--enable-exception-handling', '--enable-tail-call', '--enable-reference-types', '--enable-multivalue', '--enable-gc', '--enable-memory64', '--enable-typed-function-references', '--enable-relaxed-simd'] []
/home/taylor/binaryen/bin/wasm-opt a.wasm -o b.wasm --coalesce-locals --vacuum --roundtrip --optimize-instructions --simplify-locals-notee-nostructure --rse --optimize-instructions --pick-load-signs -fimfs=99999999 --denan --no-fuzz-oob --legalize-js-interface --enable-threads --enable-mutable-globals --enable-nontrapping-float-to-int --enable-simd --enable-bulk-memory --enable-sign-ext --enable-exception-handling --enable-tail-call --enable-reference-types --enable-multivalue --enable-gc --enable-memory64 --enable-typed-function-references --enable-relaxed-simd
post wasm size: 86315
running testcase handler: FuzzExec
/home/taylor/binaryen/bin/wasm-opt a.wasm --coalesce-locals --vacuum --roundtrip --optimize-instructions --simplify-locals-notee-nostructure --rse --optimize-instructions --pick-load-signs -fimfs=99999999 --enable-threads --enable-mutable-globals --enable-nontrapping-float-to-int --enable-simd --enable-bulk-memory --enable-sign-ext --enable-exception-handling --enable-tail-call --enable-reference-types --enable-multivalue --enable-gc --enable-memory64 --enable-typed-function-references --enable-relaxed-simd --fuzz-exec


Invocations so far:
   FuzzExec: 3
   CompareVMs: 0
   CheckDeterminism: 1
   Wasm2JS: 0
   Asyncify: 0

ITERATION: 4 seed: 9411794731008236708 size: 17971 (mean: 31663.5, stddev: 34633.2730080482) speed: 2.8066133310493058 iters/sec,  136134.0779692138 wasm_bytes/sec

randomized pass debug: 
randomized feature opts: --all-features
randomized settings (NaNs, OOB, legalize): True False True
initial contents: /home/taylor/binaryen/test/./lit/passes/once-reduction.wast
  picked submodule 7 from multi-module wast
randomized opts: --duplicate-function-elimination --generate-stack-ir --optimize-stack-ir

/home/taylor/binaryen/bin/wasm-opt input.dat -ttf -o a.wasm --no-fuzz-oob --legalize-js-interface --all-features --disable-exception-handling --disable-memory64 --disable-multivalue --strip-dwarf --initial-fuzz=initial.wat
pre wasm size: 15102
warning: no output file specified, not emitting output
warning: no output file specified, not emitting output
warning: cannot de-nan outside of function context
warning: cannot de-nan outside of function context
warning: cannot de-nan outside of function context
warning: cannot de-nan outside of function context
warning: no output file specified, not emitting output
warning: no output file specified, not emitting output
warning: no output file specified, not emitting output
warning: no passes specified, not doing any work
warning: no output file specified, not emitting output
warning: no output file specified, not emitting output
warning: no output file specified, not emitting output
  File "/home/taylor/binaryen/./scripts/fuzz_opt.py", line 1300, in <module>
    total_wasm_size += test_one(raw_input_data, given_wasm)
  File "/home/taylor/binaryen/./scripts/fuzz_opt.py", line 1105, in test_one
    testcase_handler.handle_pair(input=random_input, before_wasm='a.wasm', after_wasm='b.wasm', opts=opts + FEATURE_OPTS)
  File "/home/taylor/binaryen/./scripts/fuzz_opt.py", line 765, in handle_pair
    before = self.run_vms(before_wasm)
  File "/home/taylor/binaryen/./scripts/fuzz_opt.py", line 775, in run_vms
    vm_results[vm] = fix_output(vm.run(wasm))
  File "/home/taylor/binaryen/./scripts/fuzz_opt.py", line 636, in run
    return run_vm([shared.V8, wasm + '.js'] + shared.V8_OPTS + extra_d8_flags + ['--', wasm])
  File "/home/taylor/binaryen/./scripts/fuzz_opt.py", line 530, in run_vm
    return filter_known_issues(run(cmd))
  File "/home/taylor/binaryen/./scripts/fuzz_opt.py", line 76, in run
    print(' '.join(cmd))
/home/taylor/binaryen/bin/wasm-opt a.wasm --all-features --disable-exception-handling --disable-memory64 --disable-multivalue --strip-dwarf --print-features
['--enable-threads', '--enable-mutable-globals', '--enable-nontrapping-float-to-int', '--enable-simd', '--enable-bulk-memory', '--enable-sign-ext', '--enable-tail-call', '--enable-reference-types', '--enable-gc', '--enable-typed-function-references', '--enable-relaxed-simd'] ['--strip-dwarf']
/home/taylor/binaryen/bin/wasm-opt a.wasm -o b.wasm --duplicate-function-elimination --generate-stack-ir --optimize-stack-ir --no-fuzz-oob --legalize-js-interface --enable-threads --enable-mutable-globals --enable-nontrapping-float-to-int --enable-simd --enable-bulk-memory --enable-sign-ext --enable-tail-call --enable-reference-types --enable-gc --enable-typed-function-references --enable-relaxed-simd --strip-dwarf
post wasm size: 15102
running testcase handler: CheckDeterminism
/home/taylor/binaryen/bin/wasm-opt a.wasm -o b1.wasm --duplicate-function-elimination --generate-stack-ir --optimize-stack-ir --enable-threads --enable-mutable-globals --enable-nontrapping-float-to-int --enable-simd --enable-bulk-memory --enable-sign-ext --enable-tail-call --enable-reference-types --enable-gc --enable-typed-function-references --enable-relaxed-simd --strip-dwarf
/home/taylor/binaryen/bin/wasm-opt a.wasm -o b2.wasm --duplicate-function-elimination --generate-stack-ir --optimize-stack-ir --enable-threads --enable-mutable-globals --enable-nontrapping-float-to-int --enable-simd --enable-bulk-memory --enable-sign-ext --enable-tail-call --enable-reference-types --enable-gc --enable-typed-function-references --enable-relaxed-simd --strip-dwarf

running testcase handler: FuzzExec
/home/taylor/binaryen/bin/wasm-opt a.wasm --duplicate-function-elimination --generate-stack-ir --optimize-stack-ir --enable-threads --enable-mutable-globals --enable-nontrapping-float-to-int --enable-simd --enable-bulk-memory --enable-sign-ext --enable-tail-call --enable-reference-types --enable-gc --enable-typed-function-references --enable-relaxed-simd --strip-dwarf --fuzz-exec


Invocations so far:
   FuzzExec: 4
   CompareVMs: 0
   CheckDeterminism: 2
   Wasm2JS: 0
   Asyncify: 0

ITERATION: 5 seed: 12153351308520168699 size: 6229 (mean: 26576.6, stddev: 32604.862917055794) speed: 2.9837424145590385 iters/sec,  133804.73508393427 wasm_bytes/sec

randomized pass debug: 
randomized feature opts: --all-features --disable-threads --disable-mutable-globals --disable-nontrapping-float-to-int --disable-simd --disable-bulk-memory --disable-sign-ext --disable-exception-handling --disable-tail-call --disable-reference-types --disable-multivalue --disable-gc --disable-memory64 --disable-typed-function-references --disable-relaxed-simd
randomized settings (NaNs, OOB, legalize): False False True
initial contents: /home/taylor/binaryen/test/spec/multivalue.wast
(initial contents not valid for features, ignoring)
randomized opts: --rse --roundtrip --generate-stack-ir --code-pushing -O2 --code-folding -fimfs=99999999

/home/taylor/binaryen/bin/wasm-opt input.dat -ttf -o a.wasm --denan --no-fuzz-oob --legalize-js-interface --all-features --disable-threads --disable-mutable-globals --disable-nontrapping-float-to-int --disable-simd --disable-bulk-memory --disable-sign-ext --disable-exception-handling --disable-tail-call --disable-reference-types --disable-multivalue --disable-gc --disable-memory64 --disable-typed-function-references --disable-relaxed-simd --disable-exception-handling --disable-memory64 --disable-multivalue --strip-dwarf
pre wasm size: 4713
/home/taylor/binaryen/bin/wasm-opt a.wasm --all-features --disable-threads --disable-mutable-globals --disable-nontrapping-float-to-int --disable-simd --disable-bulk-memory --disable-sign-ext --disable-exception-handling --disable-tail-call --disable-reference-types --disable-multivalue --disable-gc --disable-memory64 --disable-typed-function-references --disable-relaxed-simd --disable-exception-handling --disable-memory64 --disable-multivalue --strip-dwarf --print-features
[] ['--strip-dwarf']
/home/taylor/binaryen/bin/wasm-opt a.wasm -o b.wasm --rse --roundtrip --generate-stack-ir --code-pushing -O2 --code-folding -fimfs=99999999 --denan --no-fuzz-oob --legalize-js-interface --strip-dwarf
post wasm size: 1557
running testcase handler: Wasm2JS
/home/taylor/binaryen/bin/wasm-opt a.wasm --legalize-js-interface -o a.wasm.temp.wasm --strip-dwarf
/home/taylor/binaryen/bin/wasm-opt a.wasm.temp.wasm -o a.wasm.temp.wasm --stub-unsupported-js --strip-dwarf
/home/taylor/binaryen/bin/wasm-opt a.wasm.temp.wasm -o b.wasm.temp.wasm --rse --roundtrip --generate-stack-ir --code-pushing -O2 --code-folding -fimfs=99999999 --strip-dwarf --strip-dwarf
/home/taylor/binaryen/bin/wasm-opt a.wasm.temp.wasm --emit-js-wrapper=/dev/stdout --strip-dwarf
/home/taylor/binaryen/bin/wasm2js a.wasm.temp.wasm --emscripten -O --deterministic --strip-dwarf
/usr/local/bin/node a.wasm.temp.wasm.js a.wasm
/home/taylor/binaryen/bin/wasm-opt b.wasm.temp.wasm --emit-js-wrapper=/dev/stdout --strip-dwarf
/home/taylor/binaryen/bin/wasm2js b.wasm.temp.wasm --emscripten -O --deterministic --strip-dwarf
/usr/local/bin/node b.wasm.temp.wasm.js a.wasm
/home/taylor/binaryen/bin/wasm-opt a.wasm.temp.wasm --fuzz-exec-before

running testcase handler: CompareVMs
[CompareVMs] running binaryen interpreter
/home/taylor/binaryen/bin/wasm-opt a.wasm --strip-dwarf --fuzz-exec-before
[CompareVMs] running d8
/home/taylor/binaryen/bin/wasm-opt a.wasm --emit-js-wrapper=a.wasm.js --strip-dwarf
!
-----------------------------------------
Exception:
-----------------------------------------
!
sequence item 0: expected str instance, NoneType found
================================================================================
You found a bug! Please report it with

  seed: 12153351308520168699

and the exact version of Binaryen you found it on, plus the exact Python
version (hopefully deterministic random numbers will be identical).

You can run that testcase again with "fuzz_opt.py 12153351308520168699"

The initial wasm file used here is saved as /home/taylor/binaryen/out/test/original.wasm

You can reduce the testcase by running this now:

||||
vvvv


/home/taylor/binaryen/bin/wasm-reduce /home/taylor/binaryen/out/test/original.wasm '--command=bash /home/taylor/binaryen/out/test/reduce.sh' -t /home/taylor/binaryen/out/test/t.wasm -w /home/taylor/binaryen/out/test/w.wasm


^^^^
||||

Make sure to verify by eye that the output says something like this:

At least one of the next two values should be 0:
  0
  1
The following value should be 1:
  1

(If it does not, then one possible issue is that the fuzzer fails to write a
valid binary. If so, you can print the output of the fuzzer's first command
(using -ttf / --translate-to-fuzz) in text form and run the reduction from that,
passing --text to the reducer.)

You can also read "/home/taylor/binaryen/out/test/reduce.sh" which has been filled out for you and includes
docs and suggestions.

After reduction, the reduced file will be in /home/taylor/binaryen/out/test/w.wasm
================================================================================
```

**Comment by kripken** [comment]
Hmm, that `sequence item 0: expected str instance, NoneType found` should have a stack trace, that would help figure this out. Does this help?
```diff
diff --git a/scripts/fuzz_opt.py b/scripts/fuzz_opt.py
index d7c82e244..8fb3ee89a 100755
--- a/scripts/fuzz_opt.py
+++ b/scripts/fuzz_opt.py
@@ -1302,6 +1302,7 @@ if __name__ == '__main__':
             print('(stopping by user request)')
             break
         except Exception as e:
+            raise e
             # print the exception manually, so that we can show our message at
             # the very end where it won't be missed
             ex_type, ex, tb = sys.exc_info()
```
Hopefully the error is something obvious given the stack trace.

**Comment by kripken** [comment]
Looks like this stalled. I'll run some fuzzing locally and land if there are no problems.

**Comment by willcohen** [comment]
Thank you!

**Pull request merged**